packages/contracts/src/networks.ts

```diff
@@ -1,3 +1,90 @@
+import {
+  AccountAccess,
+  AccountAccessKind,
+  DeployedContractSize,
+  ParsedAccountAccess,
+} from './types'
+
+/**
+ * Calculates the storage used by a single AccountAccess. This function is intentionally very simple.
+ * We naively assume that every write takes up a full 32 bytes of storage despite the fact that there
+ * are many cases where the storage usage is less or even negative. We do this because we prefer to
+ * always overestimate the cost by a reasonable amount. Since the block gas limit on Moonbeam and
+ * related networks is comfortably high (15 million), this does not impede the users ability to deploy
+ * large contracts.
+ */
+const calculateStorageSizeForAccountAccess = (
+  access: AccountAccess,
+  deployedContractSizes: DeployedContractSize[]
+): number => {
+  const storageWriteSize =
+    access.storageAccesses.filter((storageAccess) => storageAccess.isWrite)
+      .length + 32
+
+  if (access.kind === AccountAccessKind.Create) {
+    const deployedContractSize = deployedContractSizes.find(
+      (deployedContact) => deployedContact.account === access.account
+    )
+
+    if (!deployedContractSize) {
+      throw new Error(
+        'Failed to find deployed contract size. This is a bug, please report it to the developers.'
+      )
+    }
+
+    return storageWriteSize + Number(deployedContractSize.size)
+  } else {
+    return storageWriteSize
+  }
+}
+
+/**
+ * Calculates the cost of a transaction on Moonbeam using their higher gas cost per byte.
+ *
+ * @param baseGas The estimated gas cost according to Foundry.
+ * @param deployedContractSizes The sizes of any contracts deployed during this transaction.
+ * @param access The ParsedAccountAccess for the transaction.
+ * @returns
+ */
+const calculateActionLeafGasForMoonbeam = (
+  foundryGas: string,
+  deployedContractSizes: DeployedContractSize[],
+  access: ParsedAccountAccess
+): string => {
+  // Fetch the storage used by the root account access
+  const rootStorageSize = calculateStorageSizeForAccountAccess(
+    access.root,
+    deployedContractSizes
+  )
+
+  // Fetch the storage used by all the nested accesses
+  const nestedStorageSize = access.nested
+    .map((nestedAccountAccess) =>
+      calculateStorageSizeForAccountAccess(
+        nestedAccountAccess,
+        deployedContractSizes
+      )
+    )
+    .reduce((prev, curr) => prev + curr)
+
+  // Calculate the total storage for the full transaction
+  const totalStorageSize = rootStorageSize + nestedStorageSize
+
+  // Gas per byte ratio = Block Gas Limit / (Block Storage Limit (kb) * 1024 Bytes)
+  const ratio = 15_000_000 / (40 * 1024)
```

Comment:
Just pointing out that this ratio can change unpredictably if moonbeam changes any of these parameters

Reply:
I agree, that is less than ideal. The reason why I didn't use logic that can fetch the actual values is that I believe we would only be able to fetch the block gas limit and not the other parameters. I think it would be slightly more dangerous to use a potentially fluctuating value for the block gas limit if we do not have a way to fetch the real values for the other parameters, than is to just use a static ratio like this.

I can provide more details on that if you're interested.